### ISSUE TITLE
A770-EXCELLENCE-ROADMAP: make Intel Arc A770 a release-quality BitNet inference appliance

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/bitnet-rs-swarm#1894]
## Goal

Turn the current Intel Arc A770 OpenCL lane from a diagnostic QK256 contribution path into a dependable, fast, fallback-free BitNet appliance for the exact official Microsoft BitNet 2B I2_S/QK256 model.

The completed lane must let a normal operator select the exact A770 route, load and verify the supported model bundle, ask questions, reuse a resident session, inspect execution/quality/performance receipts, and receive precise failures without hidden CPU fallback.

## Current verified boundary

The lane already has substantial real evidence:

- exact selected device identity: `intel-arc-a770-opencl`, `runtime_api=opencl`;
- strict answer runs with `fallback_used=false`;
- complete model answers with all observed BitNet QK256 linears routed through A770 OpenCL;
- a repaired seeded 20-case corpus that passes on both CPU and A770;
- one remaining generated-output divergence and systematic CPU/A770 top-k logit drift;
- ninety focused Q/K/V row replays across layers 0–29;
- production-kernel disassembly and focused production replay evidence;
- the projection replay hook, currently blocked on a physical full-projection packed-row source.

The runtime is still product-hostile: it creates an OpenCL context/queue/program/kernel and allocates/uploads/reads buffers per linear invocation. Packed weights are repeatedly copied; activation quantization remains on the CPU; prefill is repeated GEMV; support-op ownership is not expressed as a complete execution plan; performance and full residency claims remain closed.

### A. Trusted hybrid A770 appliance — first product target

Required:

- official model/tokenizer/prompt identity;
- complete QK256 projection and generated-token correctness gates;
- persistent OpenCL session and compiled program cache;
- packed weights uploaded once and resident;
- reusable activation/output/scratch storage;
- optimized decode GEMV and batched prefill path;
- explicit operation/state ownership for every transformer stage;
- fallback-free exact-route receipts;
- bounded answer quality, reliability, and current-head benchmark qualification;
- normal doctor/status/ask/chat and optional exact-profile local server UX.

CPU-owned support operations are allowed only when named explicitly in the execution plan and receipt. This finish line does **not** imply full device residency.

### B. Full-device-resident A770 inference — separate promotion

Additionally requires independently proven A770 ownership for embeddings, norms, RoPE, attention scores, softmax, value mix, KV reads/writes, residuals, final norm, tied LM head, and any other support operation required by the exact model route. Full residency must not block delivery of the trusted hybrid appliance.

### Wave 0 — close the live source blocker

- #1891 — capture real full projection packed QK256 rows.
- #1893 — external source-packet handoff; land only with exact-head validation and without pretending it supplies the missing physical rows.

### Wave 1 — correctness convergence

- complete layer-0 Q/K/V projection replay;
- expand to O and FFN gate/up/down projections;
- prove one complete transformer layer;
- reconnect to focused logits, generated token IDs, the seeded corpus, and a pinned upstream reference subset when available.

### Wave 2 — resident runtime

- persistent context/queue/program/kernel cache;
- resident packed weight handles;
- reusable activation/output/scratch buffers;
- remove repeated host materialization and add device-side activation quantization where earned.

### Wave 3 — performance architecture

- optimized resident decode GEMV;
- true batched prefill GEMM;
- event-based kernel/transfer timing and phase accounting.

### Wave 4 — product qualification

- complete hybrid execution plan and receipt fragments;
- bounded quality, long-decode, repeated-session, cancellation, failure, and recovery gates;
- current-head performance/power/thermal/resource qualification and two distinct same-device history receipts;
- normal operator UX and exact-profile local service qualification.

### Wave 5 — full-residency research and promotion

- move named support operations one at a time;
- preserve token/quality parity and explicit residency proof;
- promote full-device residency only after every required operation has evidence.

### Immediate live boundary

- #1891 — real full-projection packed-row capture and bounded layer-0 Q/K/V replay.
- PR #1893 — source-packet handoff; not a replacement for #1891’s physical source.

### Correctness

- #1895 — complete Q/K/V/O and FFN projection replay plus one full transformer layer.
- #1896 — close focused logits and generated-token parity on the official model.

### Resident runtime

- #1897 — persistent device/context/queue/program/kernel cache.
- #1898 — model-bound resident packed QK256 weight handles.
- #1899 — reusable activation/output/scratch buffers and lower host materialization.
- #1900 — device I8_S activation quantization and shared projection reuse.

### Performance architecture

- #1901 — optimized resident decode QK256 GEMV.
- #1902 — true batched QK256 prefill GEMM.

### Proof, product, and promotion

- #1903 — complete planned/observed operation and state ownership receipts.
- #1904 — exact-profile answer quality, long decode, and runtime reliability.
- #1905 — doctor/status/model/ask/resident-chat/operator UX.
- #1906 — current-head benchmark, resources, power/thermals, and same-device history.
- #1907 — optional exact-profile local service qualification.
- #1909 — integrate and promote the trusted hybrid A770 route.

### Later full residency

- #1908 — move support operations/state to A770 and prove full-device-resident inference.

### CI control plane

- #1910 — cheap proof-scoped validation and explicit physical-A770 routing for all roadmap PRs.

## Recommended execution order

```text
#1893 exact-head review
→ #1891 physical source/replay
→ #1895 complete layer correctness
→ #1896 generated-token correctness

In parallel after contracts stabilize:
  #1910 CI routing
  #1903 execution-plan schema

Then:
  #1897 persistent runtime
  → #1898 resident weights
  → #1899 reusable storage
  → #1900 device quantization

Then in parallel:
  #1901 decode GEMV
  #1902 prefill GEMM

Then:
  #1904 quality/reliability
  #1905 operator UX
  #1906 benchmark/history
  → #1909 trusted-hybrid promotion
  → optional #1907 local service

Later and independently:
  #1908 full-device residency
```

## Global implementation rules

- One issue slice and one semantic objective per PR.
- Do not continue one-output-row-per-PR archaeology after the full projection source exists.
- Do not fabricate packed weights or commit model binaries.
- Do not change answer scoring or sampling to hide numerical drift.
- A770 proof must be selected-device OpenCL with `fallback_used=false`; generic OpenCL, Arc 140V, OpenVINO GPU, CUDA, NPU, or CPU cannot substitute.
- Keep CUDA, Lunar Lake, Apple, dense-SLM, server-generalization, and unrelated CI work outside A770 PRs.
- Most PRs use cheap focused validation. Physical A770 runs occur only when the issue’s evidence gate requires them.
- No speed claim before correctness, residency, route, model, quality, resource, and history gates are clean.
- No full-residency claim from a fast QK256 kernel alone.

## Codex operating order

1. Finish the currently active #1891/#1893 boundary.
2. Select the first unblocked child issue above.
3. Characterize before redesign when behavior is not already receipt-backed.
4. Land scaffolding before promotion.
5. Record the exact selected route, fallback state, proof, and not-claims in every physical receipt.
6. Close out the campaign item and generated state after each merge.

## Completion criteria

The roadmap is complete only when:

- the exact official model produces stable, bounded-correct answers through the selected A770 route;
- no claimed QK256 linear falls back to CPU;
- weights and reusable runtime resources are resident across a session;
- prefill and decode use purpose-built paths;
- operation ownership, transfers, kernel counts, timing, memory, driver, PCIe/ReBAR, power, and thermal context are explainable;
- supported commands work without undocumented environment variables;
- reliability and same-device history are proven;
- support matrices and claim ledgers derive only the exact claims earned by receipts;
- full residency remains separately and honestly classified until its own gates pass.

## Existing related work

- #1797 — broader Intel Arc runtime profiling/generalization; A770 behavior must remain exact and must not be destabilized by Arc 140V work.
- #1869 — upstream BitNet compatibility roadmap; consume its pinned reference contracts rather than recreating them here.
- #1791 — repository-wide release-stable quality gate; A770 consumes it rather than inventing a competing quality system.
